### PR TITLE
Complete-er C# Bindings

### PR DESCRIPTION
**Pull request by Speykious**
We got the basic C# bindings done. However, there are still a lot of C# bindings remaining to be done, especially if we aim at supporting *all* the Mediapipe library including all the graphs.

### CODE REVIEW

**Comment by sr229** [comment]
It might be better to merge GH-8 in this as a big PR so you can consistently test behavior while completing the interfaces.

**Comment by Speykious** [comment]
That was my idea but @adryzz wanted to be more organized.
What do you think @adryzz, do we make it into a big PR?

**Comment by sr229** [comment]
I can merge GH-8 now so it's easier.

**Comment by Speykious** [comment]
This PR is an absolute mess... I mean at least it was

**Comment by Speykious** [comment]
A possible way to force `T` to be a `Packet` would be to define the function like this:
```cs
        public P At<T, P>(string key) where P : Packet<T>
        {
            UnsafeNativeMethods.mp_SidePacket__at__PKc(MpPtr, key, out var packetPtr);

            if (packetPtr == IntPtr.Zero)
                return default(P);

            GC.KeepAlive(this);
            return (P)Activator.CreateInstance(typeof(P), packetPtr, true);
        }
```
But it introduces a redundancy where you have to specify the underlying type of the `Packet` as well as the type of the inheriting class:
```cs
Assert.IsNull(sidePacket.At<float, FloatPacket>("value"));
// instead of
Assert.IsNull(sidePacket.At<FloatPacket>("value"));
```
If it's a good solution *and* we can determine that it won't cause us any problems, maybe we'll add that in; but until then, let's keep `T` unforced like it's done currently in MediapipeUnityPlugin.

**Review by SteveGremory — APPROVED** [review]
Seems fine

**Comment by Speykious** [comment]
The PR is now **really** complete. All tests, including the `HelloWorldGraphTest.cs`, run perfectly and pass successfully.
We're very, *very* close to an actually usable library. :D

File: src/Akihabara/Framework/Packet/SidePacket.cs [hunk 1]
```diff
@@ -20,17 +20,17 @@ protected override void DeleteMpPtr()
...
         {
             UnsafeNativeMethods.mp_SidePacket__at__PKc(MpPtr, key, out var packetPtr);
 
             if (packetPtr == IntPtr.Zero)
-                return default(Packet<T>);
+                return default(T);
 
             GC.KeepAlive(this);
-            return (Packet<T>)Activator.CreateInstance(typeof(Packet<T>), packetPtr, true);
+            return (T)Activator.CreateInstance(typeof(T), packetPtr, true);
```
Comment (adryzz) [inline src/Akihabara/Framework/Packet/SidePacket.cs:33]: Is `Activator.CreateInstance()` the only way here? Reflection is a lot slower.

**Review by adryzz — COMMENTED** [review]
Looks good, just the one reflection thing.

File: src/Akihabara/Framework/Packet/SidePacket.cs (hunk 1 above)
Comment (Speykious) [reply src/Akihabara/Framework/Packet/SidePacket.cs:33]: If we decide to force `T` to be a `Packet` [like suggested here](https://github.com/vignetteapp/Akihabara/pull/9#issuecomment-920487957), maybe we can do something better like `new P(packetPtr, true)` or something. But in this case I don't know.

Comment (Speykious) [reply src/Akihabara/Framework/Packet/SidePacket.cs:33]: I'll look around to see if there are any better solutions

Comment (adryzz) [reply src/Akihabara/Framework/Packet/SidePacket.cs:33]: That would speed things up, but we need to check if `T` is always a `Packet`.

Comment (Speykious) [reply src/Akihabara/Framework/Packet/SidePacket.cs:33]: That's suggested by Homuler's To-do but yeah I guess we'd have to verify.

Comment (adryzz) [reply src/Akihabara/Framework/Packet/SidePacket.cs:33]: I'd say change it and then if we find any, we'll revert the commit after merge.

Comment (Speykious) [reply src/Akihabara/Framework/Packet/SidePacket.cs:33]: What about [adding a constraint like this](https://stackoverflow.com/a/731462) to use `new`?

Comment (adryzz) [reply src/Akihabara/Framework/Packet/SidePacket.cs:33]: Sounds good as it has reflection as a fallback.

Comment (Speykious) [reply src/Akihabara/Framework/Packet/SidePacket.cs:33]: Alright I'll do that once I get back on my computer

Comment (sr229) [reply src/Akihabara/Framework/Packet/SidePacket.cs:33]: Going to veto this: We stick with the 1-to-1 behavior, any changes can be done later.

Comment (adryzz) [reply src/Akihabara/Framework/Packet/SidePacket.cs:33]: Alright, maybe add a comment in the code linking to this thread for future reference

**Review by sr229 — APPROVED** [review]
Start the reflection fix on another PR as that one is out of scope for this. Would love benchmarks too to compare how fast the two were compared to each other

**Comment by Speykious** [comment]
Awesome!

**Comment by Transfusion** [comment]
Some thoughts on the packet issue -
Here's another place where reflection is being used (in the main `Packet.cs`)
https://github.com/homuler/MediaPipeUnityPlugin/blob/master/Packages/com.github.homuler.mediapipe/Runtime/Scripts/Framework/Packet/Packet.cs#L19-L24

**Pull request merged**